src/engine/GroupByHashMapOptimization.h

```diff
@@ -191,3 +191,34 @@ struct GroupConcatAggregationData {
 
   void reset() { currentValue_.clear(); }
 };
+
+// Data to perform SAMPLE aggregation using the HashMap optimization.
+struct SampleAggregationData {
+  std::optional<sparqlExpression::IdOrLiteralOrIri> value_ = std::nullopt;
+
+  // _____________________________________________________________________________
+  void addValue(const sparqlExpression::IdOrLiteralOrIri& value,
+                [[maybe_unused]] const sparqlExpression::EvaluationContext*) {
+    if (!value_.has_value()) {
+      value_.emplace(value);
+    }
+  }
+
+  // _____________________________________________________________________________
+  [[nodiscard]] ValueId calculateResult(LocalVocab* localVocab) const {
+    if (!value_.has_value()) {
+      return Id::makeUndefined();
+    }
+    auto valueIdResultGetter = [](ValueId id) { return id; };
+    auto stringResultGetter =
+        [localVocab](const ad_utility::triple_component::LiteralOrIri& str) {
+          auto localVocabIndex = localVocab->getIndexAndAddIfNotContained(str);
+          return ValueId::makeFromLocalVocabIndex(localVocabIndex);
+        };
+    return std::visit(ad_utility::OverloadCallOperator{valueIdResultGetter,
+                                                       stringResultGetter},
+                      value_.value());
+  }
```

Comment:
This function

(Take an `IdOrLiteralOrIri` and a `LocalVocab` and turn it into a `ValueId` is now duplicated several times across the codebase. Consider implementing it in the `SparqlExpressionGenerators` file and reuse it here.
You should then of course
1. Add unit tests for the isolated method
2. Move this code here (and all similar code) into a `.cpp` file (for calculate result we can and should definitely afford the indirection.